### PR TITLE
Kubernetes CSI - Persistent Volume Source Type

### PR DESCRIPTION
**Pull request by vladimirvivien**
**What this PR does / why we need it**:
This PR is to track the addition of new API type `CSIPersistentVolumeSource` that will be used as PersistentVolume for storage sources managed by CSI drivers. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
xref https://github.com/kubernetes/features/issues/178

**Special notes for your reviewer**:
- Implements API `PersistentVolume` type `CSIPersistentVolumeSource`
- Part of implementation for https://github.com/kubernetes/features/issues/178
- Designed at https://github.com/kubernetes/community/pull/1258

Other CSI Volume Plugin PRs:
- Plugin Mounter/Unmounter https://github.com/kubernetes/kubernetes/pull/54529
- Plugin Attacher/Detacher https://github.com/kubernetes/kubernetes/pull/55809

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -1503,6 +1506,36 @@ type LocalVolumeSource struct {
...
 
+// Represents storage that is managed by an external CSI volume driver
+type CSIPersistentVolumeSource struct {
+	// Driver is the name of the driver to use for this volume.
+	// Required.
+	Driver string
+
+	// VolumeHandle is the unique volume name returned by the CSI volume
+	// plugin’s CreateVolume to refer to the volume on all subsequent calls.
+	VolumeHandle string
```
Comment (dixudx) [inline pkg/api/types.go]: Add `// Required` for `VolumeHandle`.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1309,6 +1309,19 @@ func validateStorageOSPersistentVolumeSource(storageos *api.StorageOSPersistentV
...
+func validateCSIPersistentVolumeSource(csi *api.CSIPersistentVolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(csi.Driver) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("driver"), ""))
+	}
+
+	if len(csi.VolumeHandle) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("volumeHandle"), ""))
+	}
+
```
Comment (dixudx) [inline pkg/api/validation/validation.go]: Also need to validate `MountSecretRef` and `AttachSecretRef` if they are not nil.

Like, 
```
if csi.MountSecretRef != nil {
		if len(csi.MountSecretRef.Name) == 0 {
			allErrs = append(allErrs, field.Required(fldPath.Child("MountSecretRef", "name"), ""))
		}
		...
}
```

**Review by dixudx — COMMENTED** [review]
@vladimirvivien I didn't see `MountInterface` was implemented for `csi` in this PR.

@jsafrane @rootfs PTAL.

File: pkg/api/types.go (hunk 1 above)
Comment (vladimirvivien) [reply pkg/api/types.go]: @dixudx will add `// Required`

**Comment by vladimirvivien** [comment]
@dixudx thanks for the review.  What do you mean by `MountInterface`? (Are you referring to the volume.Mounter interface ?) Please point me where I can do it.  Thank you!

I am also working on https://github.com/kubernetes/kubernetes/pull/54529 which is the actual volume plugin impl for CSI.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (vladimirvivien) [reply pkg/api/validation/validation.go]: @dixudx ok good catch. Will look into this as well.  Thanks.

Comment (vladimirvivien) [reply pkg/api/validation/validation.go]: @dixudx I dbl checked, `MountSecretRef` is no longer part of the API type.

File: pkg/api/types.go [hunk 3]
```diff
@@ -1503,6 +1506,37 @@ type LocalVolumeSource struct {
...
+	// plugin’s CreateVolume to refer to the volume on all subsequent calls.
+	// Required.
+	VolumeHandle string
+
+	// Optional: MountSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver during NodePublish.
+	// This may be empty if no secret is required. If the secret object contains
+	// more than one secret, all secrets are passed.
+	// +optional
+	MountSecretRef *SecretReference
```
Comment (jsafrane) [inline pkg/api/types.go]: did we remove secret refs in the proposal?

Comment (vladimirvivien) [reply pkg/api/types.go]: Yes.  Master  branch does not have it : https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md#proposed-api

File: pkg/printers/internalversion/describe.go [hunk 4]
```diff
@@ -1073,6 +1073,14 @@ func printFlockerVolumeSource(flocker *api.FlockerVolumeSource, w PrefixWriter)
 		flocker.DatasetName, flocker.DatasetUUID)
 }
 
+func printCSIPersistentVolumeSource(csi *api.CSIPersistentVolumeSource, w PrefixWriter) {
+	w.Write(LEVEL_2, "Type:\tCSIVolume (a generic volume resource that represents a volume managed by an external CSI plugin)\n"+
```
[the remark this answers is not in the document]
Comment (vladimirvivien) [reply pkg/printers/internalversion/describe.go]: Added full CSI name

File: pkg/api/validation/validation_test.go [hunk 5]
```diff
@@ -1224,6 +1224,44 @@ func TestValidateGlusterfs(t *testing.T) {
...
+		errfield string
+	}{
+		{
+			name:     "missing driver name",
+			csi:      &api.CSIPersistentVolumeSource{VolumeHandle: "test-123"},
+			errtype:  field.ErrorTypeRequired,
+			errfield: "driver",
+		},
+		{
+			name:     "missing volume handle",
```
[the remark this answers is not in the document]
Comment (vladimirvivien) [reply pkg/api/validation/validation_test.go]: added couple of successful tests.

File: pkg/api/types.go [hunk 6]
```diff
@@ -391,6 +391,9 @@ type PersistentVolumeSource struct {
 	// More info: https://releases.k8s.io/HEAD/examples/volumes/storageos/README.md
 	// +optional
 	StorageOS *StorageOSPersistentVolumeSource
+	// CSI represents storage that handled by an external CSI driver
```
[the remark this answers is not in the document]
Comment (vladimirvivien) [reply pkg/api/types.go]: Agreed.  Added full name.

File: cmd/kube-controller-manager/app/plugins.go [hunk 7]
```diff
@@ -105,6 +107,7 @@ func ProbeExpandableVolumePlugins(config componentconfig.VolumeConfiguration) []
 	allPlugins = append(allPlugins, scaleio.ProbeVolumePlugins()...)
 	allPlugins = append(allPlugins, storageos.ProbeVolumePlugins()...)
 	allPlugins = append(allPlugins, fc.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, fc.ProbeVolumePlugins()...)
```
[the remark this answers is not in the document]
Comment (vladimirvivien) [reply cmd/kube-controller-manager/app/plugins.go]: fixed.

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -1503,6 +1506,37 @@ type LocalVolumeSource struct {
...
+	// plugin’s CreateVolume to refer to the volume on all subsequent calls.
+	// Required.
+	VolumeHandle string
+
+	// Optional: MountSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver during NodePublish.
+	// This may be empty if no secret is required. If the secret object contains
+	// more than one secret, all secrets are passed.
+	// +optional
+	MountSecretRef *SecretReference
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: Secrets were postponed to future release, please remove them.

Comment (jsafrane) [reply pkg/apis/core/types.go]: Hmm, github still shows that MountSecretRef is present in the first commit and at the same time it shows this review remark has been resolved / relevant code has been modified.

Comment (jsafrane) [reply pkg/apis/core/types.go]: Oh, I see, you put it into the commit with generated code. Sorry for the noise.

Still, IMO it should be squashed into the first commit for further review. Nobody looks at the generated stuff.

Comment (vladimirvivien) [reply pkg/apis/core/types.go]: @jsafrane ah ok, that commit slipped in to the generated stuff.  I will fix that.  Thanks.

**Comment by jsafrane** [comment]
lgtm from sig-storage point of view

/assign @thockin @liggitt 
for API review

**Comment by saad-ali** [comment]
/lgtm
/approve

@thockin for API review.

**Comment by jsafrane** [comment]
@thockin, can you please approve? It's blocking subsequent PRs.

File: pkg/apis/core/validation/validation.go [hunk 9]
```diff
@@ -1318,6 +1318,19 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
 	return allErrs
 }
 
+func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```
Comment (saad-ali) [inline pkg/apis/core/validation/validation.go]: Please add a feature gate check:
```GO
if !utilfeature.DefaultFeatureGate.Enabled(features.CSIPersistentVolume) {
  allErrs = append(allErrs, field.Forbidden(fldPath, "field is disabled by feature-gate CSIPersistentVolume"))
}
```

**Pull request merged**